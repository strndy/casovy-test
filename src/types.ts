export type Event = {
    Action: string;
    Time: Date;
    ISIN: string;
    Ticker: string;
    Name: string;
    NoOfShares: number;
    PriceShare: number;
    CurrencyPriceShare: string;
    ExchangeRate: string;
    Result: string;
    CurrencyResult: string;
    Total: string;
    CurrencyTotal: string;
    WithholdingTax: string;
    CurrencyWithholdingTax: string;
    StampDutyReserveTax: string;
    CurrencyStampDutyReserveTax: string;
    TransactionFee: string;
    FinraFee:string;
    Notes: string;
    ID: string;
    CurrencyTransactionFee: string;
    CurrencyFinraFee: string;
}

// A single share of a stock
export type Share = {
    Ticker: string;
    BuyDate: Date;
    BuyPrice: number;
    BuyEventId: string;
    SellPrice?: number;
    SellDate?: Date;
    SellEventId?: string;
}

export type Split = {
    execution_date: Date;
    id: string;
    split_from: number;
    split_to: number;
    ticker: string;
}
