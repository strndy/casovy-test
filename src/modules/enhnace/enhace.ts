import { Event, Split } from "../../types";

export const enhanceCSVWithSplits = (csvData: Event[], splits: Split[][]) : Event[] => {
    const enhancedData: Event[] = [...csvData];
    const splitEvents: Event[] = [];

    splits.forEach(stockSplits => {
        stockSplits.forEach(s => {
            const splitEvent: Event = {
                Action: 'Split',
                Time: s.execution_date,
                ISIN: '',
                Ticker: s.ticker,
                Name: `${s.ticker} Stock Split`,
                NoOfShares: 0,
                PriceShare: 0,
                CurrencyPriceShare: '',
                ExchangeRate: '',
                Result: '',
                CurrencyResult: '',
                Total: '',
                CurrencyTotal: '',
                WithholdingTax: '',
                CurrencyWithholdingTax: '',
                StampDutyReserveTax: '',
                CurrencyStampDutyReserveTax: '',
                TransactionFee: '',
                FinraFee: '',
                SplitFrom: s.split_from,
                SplitTo: s.split_to,
                Notes: `Stock split ${s.split_from}:${s.split_to} Id: ${s.id}`,
                ID: s.id,
                CurrencyTransactionFee: '',
                CurrencyFinraFee: ''
            };
            splitEvents.push(splitEvent);
        });
    });

    enhancedData.push(...splitEvents);
    enhancedData.sort((a, b) => a.Time.getTime() - b.Time.getTime());

    return enhancedData;
}

