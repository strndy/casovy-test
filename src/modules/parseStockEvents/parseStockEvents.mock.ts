export const mockStockEvents = [
  {
    Action: "Market buy",
    Time: new Date("2023-01-01T10:00:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 3,
    PriceShare: 150.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "",
    CurrencyResult: "USD",
    Total: "1500.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "111111",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  },
  {
    Action: "Market buy",
    Time: new Date("2023-02-01T10:00:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 1,
    PriceShare: 200.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "",
    CurrencyResult: "USD",
    Total: "1500.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "111112",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  },
  {
    Action: "Market sell",
    Time: new Date("2023-06-01T14:30:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 1,
    PriceShare: 175.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "875.00",
    CurrencyResult: "USD",
    Total: "875.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "111113",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  },
  {
    Action: "Limit sell",
    Time: new Date("2023-07-01T14:30:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 1,
    PriceShare: 185.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "875.00",
    CurrencyResult: "USD",
    Total: "875.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "111114",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  }
]

export const mockStockEventsOutOfOrder = [
  {
    Action: "Market buy",
    Time: new Date("2023-03-01T10:00:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 10,
    PriceShare: 150.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "",
    CurrencyResult: "USD",
    Total: "1500.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "12345",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  },
  {
    Action: "Market sell",
    Time: new Date("2023-01-01T14:30:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 5,
    PriceShare: 175.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "875.00",
    CurrencyResult: "USD",
    Total: "875.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "67890",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  }
]

export const mockStockEventsDuplicateID = [
  {
    Action: "Market buy",
    Time: new Date("2023-03-01T10:00:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 2,
    PriceShare: 150.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "",
    CurrencyResult: "USD",
    Total: "1500.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "12345",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  },
  {
    Action: "Market sell",
    Time: new Date("2023-01-01T14:30:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 1,
    PriceShare: 175.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "875.00",
    CurrencyResult: "USD",
    Total: "875.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "67890",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  },
  {
    Action: "Market sell",
    Time: new Date("2023-01-01T14:30:00Z"),
    ISIN: "US0378331005",
    Ticker: "AAPL",
    Name: "Apple Inc.",
    NoOfShares: 1,
    PriceShare: 175.00,
    CurrencyPriceShare: "USD",
    ExchangeRate: "1",
    Result: "875.00",
    CurrencyResult: "USD",
    Total: "875.00",
    CurrencyTotal: "USD",
    WithholdingTax: "",
    CurrencyWithholdingTax: "",
    StampDutyReserveTax: "",
    CurrencyStampDutyReserveTax: "",
    TransactionFee: "0.50",
    FinraFee: "0.10",
    Notes: "",
    ID: "67890",
    CurrencyTransactionFee: "USD",
    CurrencyFinraFee: "USD"
  }
]
