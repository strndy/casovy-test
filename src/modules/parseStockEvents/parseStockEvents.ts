import { Share, Event, Split } from "../../types";

const TIME_TEST_DAYS = 365 * 3;

// Split the stock events into individual shares
export const organizeStockPurchases = async (csvData: Event[], ignoreStocks: string[] = []): Promise<Record<string, Share[]>> => {
    const holdings: Record<string, Share[]> = {};
    // verify order of events
    let prevEvent: Event | null = null;

    // verify there are no duplicate events by ID
    verifyNoDuplicateEvents(csvData);

    csvData.forEach((event) => {
        if(ignoreStocks.includes(event.Ticker)) {
            return;
        }
        // verify order of events
        if (prevEvent && prevEvent.Time > event.Time) {
            throw new Error('Events are not in order');
        }
        prevEvent = event;
        // create collection of individual shares
        if (event.Action.endsWith('buy')) {
            addShareToHoldings(event, holdings);
        }

        // process splits
        if (event.Action.startsWith('Split')) {
            processSplit(event, holdings);
        }

        // we are selling the oldest shares by czech law
        if (event.Action.endsWith('sell')) {
            markSold(holdings, event);
        }
    });



    return holdings;
}

const verifyNoDuplicateEvents = (events: Event[]) => {
    const ids = events
        .filter(e => e.Action.startsWith('Market') || e.Action.startsWith('Limit'))
        .map(e => e.ID);
    const uniqueIds = new Set(ids);
    if (ids.length !== uniqueIds.size) {
        throw new Error('Duplicate events by ID');
    }
}

const addShareToHoldings = (event: Event, holdings: Record<string, Share[]>) => {
    for (let i = 0; i < event.NoOfShares; i++) {
        const share: Share = {
            Ticker: event.Ticker,
            BuyDate: event.Time,
            BuyPrice: event.PriceShare,
            BuyEventId: event.ID,
            Notes: event.Notes
        }
        if (!holdings[event.Ticker]) {
            holdings[event.Ticker] = [];
        }
        holdings[event.Ticker].push(share);
    }
}

const markSold = (holdings: Record<string, Share[]>, event: Event) => {
    if(!holdings[event.Ticker]) {
        console.warn(`No shares to sell for ${event.Ticker}. Fractional is not supported.`);
        return;
    }
    const notSold = holdings[event.Ticker].filter(s => !s.SellDate);
    
    for (let i = 0; i < event.NoOfShares; i++) {
        const selling = notSold[i];
        if(!selling) {
            console.warn(`No shares to sell for ${event.Ticker}. Fractional is not supported.`);
            return;
        }
        if(selling.SellDate) {
            throw new Error(`Share ${i} already sold`);
        }
        // console.log(`Marking share ${selling.BuyEventId} as sold on sell ${event.ID}`);
        selling.SellDate = new Date(event.Time);
        selling.SellPrice = event.PriceShare;
        selling.SellEventId = event.ID;
    }
}

const processSplit = (e: Event, holdings: Record<string, Share[]>) => {

    if(!holdings[e.Ticker]) {
        // no shares to split, ignore
        return;
    }
    if (e.SplitFrom > 1){
        console.error(e);
        throw new Error('Split from is greater than 1, not implemented');
    }
    // console.log(`Splitting ${e.Ticker} ${e.SplitTo}x on ${e.Time} from ${holdings[e.Ticker].length} shares`);
    const newHolding : Share[] = [];
    const sold = holdings[e.Ticker].filter(s => s.SellDate);
    const notSold = holdings[e.Ticker].filter(s => !s.SellDate);

    for (let i = 0; i < notSold.length; i++) {
        const oldShare = holdings[e.Ticker][i];
        const share = {
            // deep cloning to break references
            ...JSON.parse(JSON.stringify(oldShare)),
            BuyPrice: oldShare.BuyPrice / e.SplitTo,
            Notes: `Split ${e.SplitTo}x on ${e.Time.toISOString().split('T')[0]}, ${oldShare.Notes}`
        }
        for (let j = 0; j < e.SplitTo; j++) {
            newHolding.push(share);
        }
    }
    holdings[e.Ticker] = [
        // deep cloning to break references
        ...JSON.parse(JSON.stringify(sold)),
        ...JSON.parse(JSON.stringify(newHolding))
    ];
    
    // console.log(`Splitting to ${e.Ticker} ${e.SplitTo}x on ${e.Time.toISOString().split('T')[0]} to ${holdings[e.Ticker].length} shares`);
}

