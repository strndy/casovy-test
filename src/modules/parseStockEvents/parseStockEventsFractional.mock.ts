export const mockStockEventsFractional = [
    {
      Action: "Market buy",
      Time: new Date("2023-01-01T10:00:00Z"),
      ISIN: "US0378331005",
      Ticker: "AAPL",
      Name: "Apple Inc.",
      NoOfShares: 0.3,
      PriceShare: 150.00,
      CurrencyPriceShare: "USD",
      ExchangeRate: "1",
      Result: "",
      CurrencyResult: "USD",
      Total: "1500.00",
      CurrencyTotal: "USD",
      WithholdingTax: "",
      CurrencyWithholdingTax: "",
      StampDutyReserveTax: "",
      CurrencyStampDutyReserveTax: "",
      TransactionFee: "0.50",
      FinraFee: "0.10",
      Notes: "before split",
      ID: "111111",
      CurrencyTransactionFee: "USD",
      CurrencyFinraFee: "USD"
    },
    {
      Action: "Market buy",
      Time: new Date("2023-02-01T10:00:00Z"),
      ISIN: "US0378331005",
      Ticker: "AAPL",
      Name: "Apple Inc.",
      NoOfShares: 1.5,
      PriceShare: 200.00,
      CurrencyPriceShare: "USD",
      ExchangeRate: "1",
      Result: "",
      CurrencyResult: "USD",
      Total: "1500.00",
      CurrencyTotal: "USD",
      WithholdingTax: "",
      CurrencyWithholdingTax: "",
      StampDutyReserveTax: "",
      CurrencyStampDutyReserveTax: "",
      TransactionFee: "0.50",
      FinraFee: "0.10",
      Notes: "before split",
      ID: "111112",
      CurrencyTransactionFee: "USD",
      CurrencyFinraFee: "USD"
    },
    {
      Action: "Split",
      Time: new Date("2023-04-01T00:00:00Z"),
      ISIN: "US0378331005",
      Ticker: "AAPL",
      Name: "Apple Inc.",
      NoOfShares: undefined,
      PriceShare: 30.00,
      CurrencyPriceShare: "USD",
      ExchangeRate: "1",
      Result: "",
      CurrencyResult: "USD",
      Total: "",
      CurrencyTotal: "USD",
      WithholdingTax: "",
      CurrencyWithholdingTax: "",
      StampDutyReserveTax: "",
      CurrencyStampDutyReserveTax: "",
      TransactionFee: "",
      FinraFee: "",
      Notes: "2x stock split",
      ID: "111113",
      CurrencyTransactionFee: "USD",
      CurrencyFinraFee: "USD",
      SplitFrom: 1,
      SplitTo: 2
    },
    {
      Action: "Market buy",
      Time: new Date("2023-05-01T14:30:00Z"),
      ISIN: "US0378331005",
      Ticker: "AAPL",
      Name: "Apple Inc.",
      NoOfShares: 0.3,
      PriceShare: 200.00,
      CurrencyPriceShare: "USD",
      ExchangeRate: "1",
      Result: "",
      CurrencyResult: "USD",
      Total: "1500.00",
      CurrencyTotal: "USD",
      WithholdingTax: "",
      CurrencyWithholdingTax: "",
      StampDutyReserveTax: "",
      CurrencyStampDutyReserveTax: "",
      TransactionFee: "0.50",
      FinraFee: "0.10",
      Notes: "bought after split",
      ID: "111114",
      CurrencyTransactionFee: "USD",
      CurrencyFinraFee: "USD"
    },
    {
      Action: "Market sell",
      Time: new Date("2023-06-01T14:30:00Z"),
      ISIN: "US0378331005",
      Ticker: "AAPL",
      Name: "Apple Inc.",
      NoOfShares: 1,
      PriceShare: 175.00,
      CurrencyPriceShare: "USD",
      ExchangeRate: "1",
      Result: "875.00",
      CurrencyResult: "USD",
      Total: "875.00",
      CurrencyTotal: "USD",
      WithholdingTax: "",
      CurrencyWithholdingTax: "",
      StampDutyReserveTax: "",
      CurrencyStampDutyReserveTax: "",
      TransactionFee: "0.50",
      FinraFee: "0.10",
      Notes: "sell after split",
      ID: "111115",
      CurrencyTransactionFee: "USD",
      CurrencyFinraFee: "USD"
    },
    {
      Action: "Limit sell",
      Time: new Date("2023-07-01T14:30:00Z"),
      ISIN: "US0378331005",
      Ticker: "AAPL",
      Name: "Apple Inc.",
      NoOfShares: 1,
      PriceShare: 185.00,
      CurrencyPriceShare: "USD",
      ExchangeRate: "1",
      Result: "875.00",
      CurrencyResult: "USD",
      Total: "875.00",
      CurrencyTotal: "USD",
      WithholdingTax: "",
      CurrencyWithholdingTax: "",
      StampDutyReserveTax: "",
      CurrencyStampDutyReserveTax: "",
      TransactionFee: "0.50",
      FinraFee: "0.10",
      Notes: "sold after split",
      ID: "111116",
      CurrencyTransactionFee: "USD",
      CurrencyFinraFee: "USD"
    }
  ]