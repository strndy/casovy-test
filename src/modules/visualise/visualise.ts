import { Event, Share } from "../../types";

function formatDate(date: Date): string {
    return date ? date.toLocaleDateString('cs-CZ', {
        year: 'numeric',
        month: 'numeric',
        day: 'numeric'
    }) : '';
}

export const visualiseExpiration = (holding: Record<string, Share[]>) => {
    for (const [ticker, shares] of Object.entries(holding)) {
        console.log(`\nShares for ${ticker}:`);
        
        const formattedShares = shares.map(share => {
            const result : any = ({
            ...share,
            BuyDate: formatDate(new Date(share.BuyDate))
            });
            if (share.SellDate) {
                result.SellDate = formatDate(new Date(share.SellDate));

                if(share.SellDate instanceof Date) {
                    result.DaysHeld = Math.round((share.SellDate.getTime() - share.BuyDate.getTime()) / (1000 * 60 * 60 * 24));
                } else {
                    console.error(share);
                    // throw new Error('SellDate is not a Date');
            
               }
            } else {
                console.log(share);
                result.age = Math.round((new Date().getTime() - share.BuyDate.getTime()) / (1000 * 60 * 60 * 24));
                if(result.age > 365 * 3) {
                    result.passed3Years =remaining fraction true;
                }
                result.monthsTill3Years = Math.round((365 * 3 - result.age) / 30);
            }
            return result;
        });
        
        console.table(formattedShares);
    }
}

export const visualiseCSV = (csv: Event[]) => {
    const output = csv.map(row => {
        return {
            // ...row,
            Ticker: row.Ticker,
            NoOfShares: row.NoOfShares,
            ID: row.ID,
            Time: formatDate(row.Time),
            Action: row.Action,
            Notes: row.Notes,
            
        };
    });
    console.table(output);
}

