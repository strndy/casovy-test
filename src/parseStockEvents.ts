import { readCSV } from "./modules/csv";
import { Holding, StockEvent } from "./types";

const TIME_TEST_DAYS = 365 * 3;

const holding: Record<string, Record<string, Holding[]>> = {};

export const parseStockEvents = async (csvData: StockEvent[]) => {
    const queue: Record<string, StockEvent[]> = {};

    // group by ticker
    csvData.forEach((event) => {
        if (!queue[event.Ticker]) {
            queue[event.Ticker] = [];
        }
        queue[event.Ticker].push(event);
    });

    console.log(queue);

}

