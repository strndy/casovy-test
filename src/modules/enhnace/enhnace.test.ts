import { Split } from "../../types";
import { mockStockEvents } from "../parseStockEvents/parseStockEvents.mock";
import { visualiseCSV, visualiseExpiration } from "../visualise/visualise";
import { enhanceCSVWithSplits } from "./enhace";

const mockSplits : Split[][] = [
    [
        {
            id: '1',
            ticker: 'AAPL',
            split_from: 1,
            split_to: 3,
            execution_date: new Date('2023-01-15')
        }
    ]
    
];

describe('enhance', () => {
    it('should enhance csv with splits', () => { 
        
        const enhanced = enhanceCSVWithSplits(mockStockEvents, mockSplits);
        visualiseCSV(enhanced);
        expect(enhanced).toBeDefined();
        expect(enhanced).toMatchSnapshot();
    });
});
