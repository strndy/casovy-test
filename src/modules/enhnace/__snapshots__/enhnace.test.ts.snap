// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhance should enhance csv with splits 1`] = `
[
  {
    "Action": "Market buy",
    "CurrencyFinraFee": "USD",
    "CurrencyPriceShare": "USD",
    "CurrencyResult": "USD",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "USD",
    "CurrencyTransactionFee": "USD",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "1",
    "FinraFee": "0.10",
    "ID": "111111",
    "ISIN": "US0378331005",
    "Name": "Apple Inc.",
    "NoOfShares": 1,
    "Notes": "before split",
    "PriceShare": 150,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-01-01T10:00:00.000Z,
    "Total": "1500.00",
    "TransactionFee": "0.50",
    "WithholdingTax": "",
  },
  {
    "Action": "Split",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "",
    "CurrencyResult": "",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "",
    "FinraFee": "",
    "ID": "1",
    "ISIN": "",
    "Name": "AAPL Stock Split",
    "NoOfShares": 0,
    "Notes": "Stock split 1:3 Id: 1",
    "PriceShare": 0,
    "Result": "",
    "SplitFrom": 1,
    "SplitTo": 3,
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-01-15T00:00:00.000Z,
    "Total": "",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Market buy",
    "CurrencyFinraFee": "USD",
    "CurrencyPriceShare": "USD",
    "CurrencyResult": "USD",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "USD",
    "CurrencyTransactionFee": "USD",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "1",
    "FinraFee": "0.10",
    "ID": "111112",
    "ISIN": "US0378331005",
    "Name": "Apple Inc.",
    "NoOfShares": 1,
    "Notes": "before split",
    "PriceShare": 200,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-02-01T10:00:00.000Z,
    "Total": "1500.00",
    "TransactionFee": "0.50",
    "WithholdingTax": "",
  },
  {
    "Action": "Split",
    "CurrencyFinraFee": "USD",
    "CurrencyPriceShare": "USD",
    "CurrencyResult": "USD",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "USD",
    "CurrencyTransactionFee": "USD",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "1",
    "FinraFee": "",
    "ID": "111113",
    "ISIN": "US0378331005",
    "Name": "Apple Inc.",
    "NoOfShares": undefined,
    "Notes": "2x stock split",
    "PriceShare": 30,
    "Result": "",
    "SplitFrom": 1,
    "SplitTo": 2,
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-04-01T00:00:00.000Z,
    "Total": "",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Market buy",
    "CurrencyFinraFee": "USD",
    "CurrencyPriceShare": "USD",
    "CurrencyResult": "USD",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "USD",
    "CurrencyTransactionFee": "USD",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "1",
    "FinraFee": "0.10",
    "ID": "111114",
    "ISIN": "US0378331005",
    "Name": "Apple Inc.",
    "NoOfShares": 1,
    "Notes": "bought after split",
    "PriceShare": 200,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-05-01T14:30:00.000Z,
    "Total": "1500.00",
    "TransactionFee": "0.50",
    "WithholdingTax": "",
  },
  {
    "Action": "Market sell",
    "CurrencyFinraFee": "USD",
    "CurrencyPriceShare": "USD",
    "CurrencyResult": "USD",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "USD",
    "CurrencyTransactionFee": "USD",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "1",
    "FinraFee": "0.10",
    "ID": "111115",
    "ISIN": "US0378331005",
    "Name": "Apple Inc.",
    "NoOfShares": 1,
    "Notes": "sell after split",
    "PriceShare": 175,
    "Result": "875.00",
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-06-01T14:30:00.000Z,
    "Total": "875.00",
    "TransactionFee": "0.50",
    "WithholdingTax": "",
  },
  {
    "Action": "Limit sell",
    "CurrencyFinraFee": "USD",
    "CurrencyPriceShare": "USD",
    "CurrencyResult": "USD",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "USD",
    "CurrencyTransactionFee": "USD",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "1",
    "FinraFee": "0.10",
    "ID": "111116",
    "ISIN": "US0378331005",
    "Name": "Apple Inc.",
    "NoOfShares": 1,
    "Notes": "sold after split",
    "PriceShare": 185,
    "Result": "875.00",
    "StampDutyReserveTax": "",
    "Ticker": "AAPL",
    "Time": 2023-07-01T14:30:00.000Z,
    "Total": "875.00",
    "TransactionFee": "0.50",
    "WithholdingTax": "",
  },
]
`;
